[red-knot] Document 'public type of undeclared symbols' behavior (#16096)
astral-sh/ruff: crates/red_knot_python_semantic/resources/mdtest/doc/public_type_undeclared_symbols.md

# Public type of undeclared symbols

## Summary

One major deviation from the behavior of existing Python type checkers is our handling of 'public'
types for undeclared symbols. This is best illustrated with an example:

```py
class Wrapper:
    value = None

wrapper = Wrapper()

reveal_type(wrapper.value)  # revealed: Unknown | None

wrapper.value = 1
```

> Mypy and Pyright both infer a type of `None` for the type of `wrapper.value`. Consequently, both
> tools raise an error when trying to assign `1` to `wrapper.value`. But there is nothing wrong with
> this program. Raising an error here violates the [gradual guarantee] which states that *"Removing
> type annotations (making the program more dynamic) should not result in additional static type
> errors."*: If `value` were annotated with `int | None` here, Mypy and Pyright would not raise any
> errors.

Reviewer: nit: "raising an error" to me makes it sound to me like we'll panic; I prefer to talk about "emitting an error"/"issuing a diagnostic".

(This might just be because I'm used to the Python context where there's a `raise` keyword, though

)
```suggestion
Mypy and Pyright both infer a type of `None` for the type of `wrapper.value`. Consequently, both
tools emit an error when trying to assign `1` to `wrapper.value`. But there is nothing wrong with
this program. Emitting an error here violates the [gradual guarantee] which states that *"Removing
type annotations (making the program more dynamic) should not result in additional static type
errors."*: If `value` were annotated with `int | None` here, Mypy and Pyright would not emit any
errors.
```

By inferring `Unknown | None` instead, we allow arbitrary values to be assigned to `wrapper.value`.
This is a deliberate choice to prevent false positive errors on untyped code.

More generally, we infer `Unknown | T_inferred` for undeclared symbols, where `T_inferred` is the
inferred type of the right-hand side of the assignment. This gradual type represents an *unknown*
fully-static type that is *at least as large as* `T_inferred`. It accurately describes our static
knowlege about this type. In the example above, we don't know what values `wrapper.value` could
possibly contain, but we *do know* that `None` is a possibility. This allows us to catch errors
where `wrapper.value` is used in a way that is incompatible with `None`:

```py
def f(w: Wrapper) -> None:
    # This is fine
    x: int | None = w.value

    # This is not okay because `w.value` could be `None`. We therefore raise the following
    # error: "Object of type `Unknown | None` is not assignable to `int`"

>     y: int = w.value

Reviewer: You could also use this as an additional example, which might be slightly more "realistic" since users often leave local variables unannotated:

```suggestion
    y: int = w.value

    z = w.value + 42  # error
```

Author: That's similar to what I wanted to demonstrate here at first, but unfortunately, we don't detect that `None + 42` is an error, yet.

Reviewer: ah, we should probably fix that 🙃

Author: I really wanted this document to be user-facing, so I would prefer if it were free of TODOs. I now used a `chr(w.value)` call to demonstrate that we would emit a *"`Unknown | None` cannot be assigned to parameter 1 (`i`) of function `chr`; expected type `int`"* diagnostic.

I used `chr` (as opposed to something more prominent) because it is one of very few functions in `builtins` that "just takes an `int`" as a parameter, and not some complicated generic protocol :upside_down_face:

Reviewer: > I used `chr` (as opposed to something more prominent) because it is one of very few functions in `builtins` that "just takes an `int`" as a parameter, and not some complicated generic protocol 🙃

Oh no, that... Seems like it might be a typeshed bug 😆

```pycon
>>> class NotAnInt:
...   def __index__(self): return 42
...   
>>> chr(NotAnInt())
'*'
```

I can keep quiet about it until red-knot's implemented generics and protocols, though 😉

Reviewer 2: Couldn't you just as well define your own function that "just takes an int" instead of needing to use something from builtins?

Author: > Couldn't you just as well define your own function that "just takes an int"

Yes, I did that now. I wanted to keep it short, but I shouldn't trade brevity for clarity.

> Seems like it might be a typeshed bug 😆

Oh :smile:

https://github.com/python/typeshed/pull/13494

```

## False negatives

> In the first example, we demonstrated how our behavior prevents false positives. Howeer, it can also

Reviewer: ```suggestion
In the first example, we demonstrated how our behavior prevents false positives. However, it can also
```

prevent false negatives. The following example contains a bug, but Mypy and Pyright do not catch it.

> To make this a bit more realistic, imagine that `OptionalInt` is important from an external, untyped

Reviewer: ```suggestion
To make this a bit more realistic, imagine that `OptionalInt` is imported from an external, untyped
```

external, untyped module:

`optional_int.py`:

```py
class OptionalInt:
    value = 10

def reset(o):
    o.value = None
```

It is then used like this:

```py
from typing_extensions import assert_type
from optional_int import OptionalInt, reset

o = OptionalInt()

reset(o)  # Oh no...

# This assertion is incorrect, but Mypy and Pyright do not catch it. We raise the following
# error: "Actual type `Unknown | Literal[10]` is not the same as asserted type `int`"
assert_type(o.value, int)

print(o.value // 2)  # Runtime error!
```

To be fair, we only catch this due to the `assert_type` call. But the type of
`Unknown | Literal[10]` for `o.value` reflects more accurately what the possible values of `o.value`
are.

## Stricter behavior

Users can always opt in to stricter behavior by adding type annotations. For the `OptionalInt`
class, this would probably be:

```py
class OptionalInt:
    value: int | None = 10

# The following public type is now
# revealed: int | None
reveal_type(OptionalInt.value)
```

## What is meant by 'public' type?

> We refer to the 'public' type of a symbol if we assign a type to the *use* of a symbol from another
> scope. *Within* the same scope, we use a narrower type of `T_inferred` for undeclared symbols, as
> there is no way that the symbol could have been reassigned. For example:

Reviewer: ```suggestion
Red-knot models all symbols as having two types: a "public type" and a "local type". The "local type"
is the type the symbol has from the perspective of code that exists in the same scope the symbol
was originally defined in. We try to infer as precise as possible a type for the local type, as we
can statically enumerate all assignments to the symbol within that scope. The "public type" of
a symbol, on the other hand, is the type the symbol has from the perspective of code in external scopes. For example:
```

Reviewer: > Hm. This might be how it looks like from the perspective of a user, but I'm not sure if that's the most helpful mental model. I think it might be better to think in terms of the inferred type and the declared type, and then to understand how those are used to construct "local" and "public" types ([`main`/crates/red_knot_python_semantic/resources/mdtest/boundness_declaredness/public.md](https://github.com/astral-sh/ruff/blob/main/crates/red_knot_python_semantic/resources/mdtest/boundness_declaredness/public.md?rgh-link-date=2025-02-11T12%3A28%3A07Z))?

Heh. And I think I'll throw that right back at you and say that, while this mental model is the most accurate in terms of our _implementation_ (and therefore probably a better mental model for us to have), I'm not sure it's so helpful for users!

> I think the suggested description also seems to imply that there are only two types for each symbol. The public and the local type. In reality, it really depends on the _use_ of a symbol. Two "local" uses of `x` could see completely different types due to narrowing, for example.

Thanks, this is a great point.

What about something like this?

```suggestion
Red-knot applies different semantics depending on whether a symbol is accessed from the same scope
in which it was originally defined, or whether it is accessed from an external scope. External
scopes will see the symbol's "public type", which has been discussed above. But within the same
scope the symbol was defined in, we use a narrower type of `T_inferred` for undeclared symbols.
This is because, from the perspective of this scope, there is no way that the value of the symbol
could have been reassigned from external scopes. For example:
```

Author: > Red-knot

I see what you did there. You thought I wouldn't notice.

    # error: "Object of type `Unknown | None` is not assignable to `int`"
    y: int = w.value
```

## False negatives

In the first example, we demonstrated how our behavior prevents false positives. However, it can
also prevent false negatives. The following example contains a bug, but Mypy and Pyright do not
catch it. To make this a bit more realistic, imagine that `OptionalInt` is imported from an
external, untyped module:

`optional_int.py`:

```py
class OptionalInt:
    value = 10

def reset(o):
    o.value = None
```

It is then used like this:

```py
from typing_extensions import assert_type
from optional_int import OptionalInt, reset

o = OptionalInt()

reset(o)  # Oh no...

# This assertion is incorrect, but Mypy and Pyright do not catch it. We raise the following
# error: "Actual type `Unknown | Literal[10]` is not the same as asserted type `int`"
assert_type(o.value, int)

print(o.value // 2)  # Runtime error!
```

To be fair, we only catch this due to the `assert_type` call. But the type of
`Unknown | Literal[10]` for `o.value` reflects more accurately what the possible values of `o.value`

> are.

Author: Yes, fully agreed; thanks for the suggestion with the dynamic-type-detection mode. I rephrased this section now. Please let me know if I'm using "unsoundness" incorrectly here, or if it sounds overly dramatic.

## Stricter behavior

Users can always opt in to stricter behavior by adding type annotations. For the `OptionalInt`
class, this would probably be:

```py
class OptionalInt:
    value: int | None = 10

# The following public type is now
# revealed: int | None

> reveal_type(OptionalInt.value)

Reviewer 2: Instead of (or in addition to) using a `reveal_type` here, we could explicitly demonstrate an assignment of some incompatible type failing

    v: int | None = w.value

    # This function call is incorrect, because `w.value` could be `None`. We therefore emit the following
    # error: "`Unknown | None` cannot be assigned to parameter 1 (`i`) of function `chr`; expected type `int`"
    c = chr(w.value)
```

## Explicit lack of knowledge

The following example demonstrates how Mypy and Pyright's type inference of fully-static types in
these situations can lead to false-negatives, even though everything appears to be (statically)
typed. To make this a bit more realistic, imagine that `OptionalInt` is imported from an external,
untyped module:

`optional_int.py`:

```py
class OptionalInt:
    value = 10

def reset(o):
    o.value = None
```

It is then used like this:

```py
from optional_int import OptionalInt, reset

o = OptionalInt()
reset(o)  # Oh no...

# Mypy and Pyright infer a fully-static type of `int` here, which appears to make the
# subsequent division operation safe -- but it is not. We infer the following type:
reveal_type(o.value)  # revealed: Unknown | Literal[10]

print(o.value // 2)  # Runtime error!
```

We do not catch this mistake either, but we accurately reflect our lack of knowledge about

> `o.value`. Together with a possible future type-checker mode that would detect the prevalance of

Reviewer 2: ```suggestion
`o.value`. Together with a possible future type-checker mode that would detect the prevalence of
```
